In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from IPython.display import clear_output

In [2]:
test = pd.read_csv('../data/test.csv', usecols=['ciudad', 'provincia', 'lat', 'lng'])

In [3]:
test.head(2)

ciudad         provincia        lat        lng
0  Miguel Hidalgo  Distrito Federal  19.408668 -99.246767
1          Mérida           Yucatán  21.032480 -89.592424

# Se verifica si hay coordenadas extrañas.

In [4]:
test.describe()

lat           lng
count  29305.000000  29305.000000
mean      20.713830    -99.561640
std        3.102425      9.400879
min      -33.541470   -121.872408
25%       19.360691   -100.971430
50%       19.546876    -99.241648
75%       20.743229    -99.136622
max       84.414160    105.442080

##### Hay valores que pueden estar mezclados.
Para México la latitud tiene que ser positiva (Norte) y la longitud negativa (Oeste).

Las coordenadas extremas que enmarcan el territorio mexicano son:

**Norte:** 32° 43´ 06´´ = 32.718333° latitud norte, en la frontera con los Estados Unidos de América.

**Sur:** 14° 32´ 27´´ = 14.540833° latitud norte, frontera con Guatemala.

**Este:** 86° 42´ 36´´ = -86.710000° longitud oeste, en el extremo suroeste de la Isla Mujeres.

**Oeste:** 118° 27´ 24´´ = -118.456667° longitud oeste, en el Océano Pacífico.


In [5]:
test.loc[(test['lat'] < 14) | \
          (test['lat'] > 33) | \
          (test['lng'] > -85) | \
          (test['lng'] < -119)].count()

ciudad       145
provincia    145
lat          145
lng          145
dtype: int64

##### Hay 145 filas con coordenas inválidas. Se rellenan con np.NaN.

In [6]:
indice = test.loc[(test['lat'] < 14) | \
                   (test['lat'] > 33) | \
                   (test['lng'] > -85) | \
                   (test['lng'] < -119)].index

In [7]:
%%time
i = 1
for var in indice:
    clear_output(wait=True)
    test.loc[var, 'lat'] = np.NaN
    test.loc[var, 'lng'] = np.NaN
    print(i, 'de', indice.size)
    i = i + 1

145 de 145
CPU times: user 432 ms, sys: 74.8 ms, total: 506 ms
Wall time: 431 ms


##### Datos nulos por columna.

In [8]:
test.isnull().sum()

ciudad          83
provincia       42
lat          30840
lng          30840
dtype: int64

##### Datos nulos para laitud y longitud simultáneamente, con "**AND**" .

In [9]:
test.loc[(test['lat'].isnull()) & (test['lng'].isnull())].isnull().sum()

ciudad          51
provincia       31
lat          30840
lng          30840
dtype: int64

##### Datos nulos para laitud y longitud simultáneamente, con "**OR**" inclusivo .

In [10]:
test.loc[(test['lat'].isnull()) | (test['lng'].isnull())].isnull().sum()

ciudad          51
provincia       31
lat          30840
lng          30840
dtype: int64

##### Se ve que coinciden los nulos para ambas coordenadas simultáneamente.

# Se exporta el train sin coordenadas inválidas

In [11]:
test.to_csv('../data/aux_minus_bad_coord_test.csv', index=False)

In [12]:
prueba = pd.read_csv('../data/aux_minus_bad_coord_test.csv')
prueba.head(5)

ciudad         provincia        lat        lng
0      Miguel Hidalgo  Distrito Federal  19.408668 -99.246767
1              Mérida           Yucatán  21.032480 -89.592424
2            Coyoacán  Distrito Federal  19.332829 -99.152913
3  Acapulco de Juárez          Guerrero  16.860487 -99.878383
4           Tultitlán    Edo. de México  19.640482 -99.127273

In [13]:
prueba.shape

(60000, 4)